Flake support

danielfullmer opened the pull request:
Initial work on making a robotnix "flake".  

Example usage:

```
$ mkdir flake-test
$ cd flake-test
$ nix flake init -t github:danielfullmer/robotnix/flake
$ # Edit flake.nix in current directory
$ nix build
```

This should build an `img` zip which could be flashed via fastboot.

---

This is all just a WIP preview, shared here in case anyone has strong feelings about the subject.

I am (like many) new to flakes.  I specifically wonder what the right interface for a robotnix flake should be, if not something like `robotnixSystem`. I definitely would like to have a convenient way to support multiple robotnix configurations inside a single flake, and the current example `defaultPackage` picks a particular output (`img`) of the specific `dailydriver` configuration, which isn't exactly very flexible.  NixOS has the entire `nixos-rebuild` script to provide a more convenient interface around flakes containing multiple NixOS configurations, but that potentially seems like overkill.

There remain some things to investigate, including adding support for `checks` / `hydraJobs`.  IIRC there was also a talk at NixCon that mentioned some proposed standard for docs to be generated via flakes?

hmenke reviewed (commented):
Could you merge this into the main branch maybe?  In the current form this supports "dual-stack Nix", i.e. either channels or flakes, so existing configuration will continue to work.  It also makes it easier for users (probably only myself) to migrate.

danielfullmer commented on template/flake.nix:
```diff
@@ -0,0 +1,28 @@
+  };
+
+  outputs = { self, robotnix }: {
+    defaultPackage.x86_64-linux = self.robotnixConfigurations."dailydriver".img;
```
Added small comment noting this in the template

danielfullmer commented on flake.nix:
```diff
@@ -0,0 +1,33 @@
+      description = "A basic robotnix configuration";
+    };
+
+    nixosModule = import ./nixos;
```
Good suggestion, thanks!

danielfullmer commented:
@hmenke Thanks for the review! I agree, it's worthwhile to merge this into master, even if it's a bit early. It should be functional, and at the very least, it's nice to have a single place where the pinned `nixpkgs` and `android-nixpkgs` can be updated.  I also added a very basic test of the flake template functionality.  Using `nix-build` will remain the recommended path for now, but I'll try not to unintentionally break the flake build path in the future, and it's nice to have you as a user of the flake functionality. 

For reference, I've switched my personal NixOS configurations to use flakes, and now use the attestation-server via this flake: https://github.com/danielfullmer/nixos-config/commit/d473bf092293a8f2cfbff5df7d7fff7bdcc49cd3
